clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/agent/TaskHealthCachingAgent.java

```diff
@@ -138,72 +143,109 @@ public TaskHealthCachingAgent(
       Task task,
       String serviceName,
       Service loadBalancerService,
       TaskDefinition taskDefinition) {
 
+    List<TaskHealth> taskHealthList = new LinkedList<>();
+
     if (taskDefinition == null) {
-      return null;
+      log.debug("Provided task definition is null.");
+      return taskHealthList;
     }
 
     List<LoadBalancer> loadBalancers = loadBalancerService.getLoadBalancers();
+    log.debug("LoadBalancerService found {} load balancers.", loadBalancers.size());
 
     for (LoadBalancer loadBalancer : loadBalancers) {
       if (loadBalancer.getTargetGroupArn() == null) {
+        log.debug("LoadBalancer does not contain a target group arn.");
+        continue;
+      }
+
+      Optional<Integer> targetGroupPort =
+          getTargetGroupContainerPort(
+              taskDefinition.getContainerDefinitions(), loadBalancer.getContainerPort());
+      if (!targetGroupPort.isPresent()) {
+        log.debug(
+            "Container does not contain a port mapping with load balanced container port: {}.",
+            loadBalancer.getContainerPort());
         continue;
       }
 
       NetworkInterface networkInterface = task.getContainers().get(0).getNetworkInterfaces().get(0);
+
       DescribeTargetHealthResult describeTargetHealthResult =
           amazonloadBalancing.describeTargetHealth(
               new DescribeTargetHealthRequest()
                   .withTargetGroupArn(loadBalancer.getTargetGroupArn())
                   .withTargets(
                       new TargetDescription()
                           .withId(networkInterface.getPrivateIpv4Address())
-                          .withPort(
-                              taskDefinition
-                                  .getContainerDefinitions()
-                                  .get(0)
-                                  .getPortMappings()
-                                  .get(0)
-                                  .getContainerPort())));
+                          .withPort(targetGroupPort.get())));
 
       if (describeTargetHealthResult.getTargetHealthDescriptions().isEmpty()) {
+        log.debug("Target health description is empty");
         evictStaleData(task, loadBalancerService);
         continue;
       }
 
+      log.debug(
+          "Target health description is not empty and has a size of {}",
+          describeTargetHealthResult.getTargetHealthDescriptions().size());
       TargetHealthDescription healthDescription =
           describeTargetHealthResult.getTargetHealthDescriptions().get(0);
 
       TaskHealth taskHealth = makeTaskHealth(task, serviceName, healthDescription);
-      return taskHealth;
+      taskHealthList.add(taskHealth);
     }
-    return null;
+    return taskHealthList;
+  }
+
+  private Optional<Integer> getTargetGroupContainerPort(
+      List<ContainerDefinition> containerDefinitions, Integer containerPort) {
+    for (ContainerDefinition containerDefinition : containerDefinitions) {
```

Comment:
could we use `.stream` & `.filter` here to make this more concise? ([example](https://github.com/spinnaker/clouddriver/blob/master/clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/deploy/ops/CreateServerGroupAtomicOperation.java#L339))

Reply:
We could, but I want it to exit as soon as one is found, vs returning everything that matches the criteria.